### ISSUE TITLE
YOLO11 GPU and CPU inference results are different

### ISSUE THREAD

**Issue by wenwu2021** [ultralytics/ultralytics#20682]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

YOLO11 GPU and CPU inference results are different.
Pycharm GPU-seg

![Image](https://github.com/user-attachments/assets/bafd8ab4-0098-42c9-a24c-2c48d4d14f28)

Pycharm cpu-seg

![Image](https://github.com/user-attachments/assets/0851e0c0-5237-44f1-82df-3d46987687be)

Pycharm GPU-pose

![Image](https://github.com/user-attachments/assets/5beee734-89dc-42d0-a464-94bedd46ff19)

Pycharm cpu-pose

![Image](https://github.com/user-attachments/assets/d6542544-50be-4bc0-9176-6b8068d70e52)

Yolo11 Onnx GPU-seg c++ 
 person:0.890330
 person:0.890146
 person:0.846505
 bus:0.820962
Yolo11 Onnx cpu-seg  c++ 
 person:0.890330
 person:0.890146
person:0.846505
 bus:0.820962

def predict():
    model = YOLO('yolo11n-seg.pt')
    source = r'bus1.png'  
    model.predict(source, imgsz=(640, 640), save=True,device="cpu")
ultralytics-8.3.126

blas                      1.0                         mkl    conda-forge
brotli                    1.1.0                h2466b09_2    conda-forge
brotli-bin                1.1.0                h2466b09_2    conda-forge
brotli-python             1.1.0           py310h9e98ed7_2    conda-forge
bzip2                     1.0.8                h2bbff1b_6
ca-certificates           2025.4.26            h4c7d964_0    conda-forge
cairo                     1.16.0               hc68a040_5
certifi                   2025.1.31          pyhd8ed1ab_0    conda-forge
cffi                      1.17.1          py310ha8f682b_0    conda-forge
charset-normalizer        3.4.2              pyhd8ed1ab_0    conda-forge
colorama                  0.4.6              pyhd8ed1ab_1    conda-forge
contourpy                 1.3.2           py310hc19bc0b_0    conda-forge
cuda-cccl                 12.9.27                       0    nvidia
cuda-cccl_win-64          12.9.27                       0    nvidia
cuda-cudart               11.8.89                       0    nvidia
cuda-cudart-dev           11.8.89                       0    nvidia
cuda-cupti                11.8.87                       0    nvidia
cuda-libraries            11.8.0                        0    nvidia
cuda-libraries-dev        11.8.0                        0    nvidia
cuda-nvrtc                11.8.89                       0    nvidia
cuda-nvrtc-dev            11.8.89                       0    nvidia
cuda-nvtx                 11.8.86                       0    nvidia
cuda-profiler-api         12.9.19                       0    nvidia
cuda-runtime              11.8.0                        0    nvidia
cuda-version              12.9                          3    nvidia
cycler                    0.12.1             pyhd8ed1ab_1    conda-forge
eigen                     3.4.0                h91493d7_0    conda-forge
exceptiongroup            1.2.2                    pypi_0    pypi
expat                     2.7.0                he0c23c2_0    conda-forge
filelock                  3.18.0             pyhd8ed1ab_0    conda-forge
fontconfig                2.14.1               hb33846d_3
fonttools                 4.57.0          py310h38315fa_0    conda-forge
freeglut                  3.4.0                hd77b12b_0
freetype                  2.13.3               h0620614_0
fsspec                    2025.3.2                 pypi_0    pypi
glib                      2.78.4               hd77b12b_0
glib-tools                2.78.4               hd77b12b_0
graphite2                 1.3.14               hd77b12b_1
gst-plugins-base          1.24.12              h1d251b9_0
gstreamer                 1.24.12              h9a7065a_0
gstreamer-orc             0.4.41               h1f81b68_0    conda-forge
h2                        4.2.0              pyhd8ed1ab_0    conda-forge
harfbuzz                  10.2.0               hf458a27_0
hdf5                      1.14.5               ha36df97_2
hpack                     4.1.0              pyhd8ed1ab_0    conda-forge
hyperframe                6.1.0              pyhd8ed1ab_0    conda-forge
icc_rt                    2022.1.0             h6049295_2
icu                       73.2                 h63175ca_0    conda-forge
idna                      3.10               pyhd8ed1ab_1    conda-forge
iniconfig                 2.1.0                    pypi_0    pypi
intel-openmp              2025.1.0           h57928b3_980    conda-forge
jinja2                    3.1.6              pyhd8ed1ab_0    conda-forge
jpeg                      9e                   h8ffe710_2    conda-forge
kiwisolver                1.4.7           py310hc19bc0b_0    conda-forge
krb5                      1.21.3               hdf4eb48_0    conda-forge
lcms2                     2.16                 h62be587_1
lerc                      4.0.0                h6470a55_1    conda-forge
libabseil                 20250127.0      cxx17_h4eb7d71_0    conda-forge
libblas                   3.9.0           1_h8933c1f_netlib    conda-forge
libbrotlicommon           1.1.0                h2466b09_2    conda-forge
libbrotlidec              1.1.0                h2466b09_2    conda-forge
libbrotlienc              1.1.0                h2466b09_2    conda-forge
libcblas                  3.9.0           12_hb3dda5d_netlib    conda-forge
libcublas                 11.11.3.6                     0    nvidia
libcublas-dev             11.11.3.6                     0    nvidia
libcufft                  10.9.0.58                     0    nvidia
libcufft-dev              10.9.0.58                     0    nvidia
libcurand                 10.3.10.19                    0    nvidia
libcurand-dev             10.3.10.19                    0    nvidia
libcusolver               11.4.1.48                     0    nvidia
libcusolver-dev           11.4.1.48                     0    nvidia
libcusparse               11.7.5.86                     0    nvidia
libcusparse-dev           11.7.5.86                     0    nvidia
libdeflate                1.22                 h2466b09_0    conda-forge
libexpat                  2.7.0                he0c23c2_0    conda-forge
libffi                    3.4.4                hd77b12b_1
libglib                   2.78.4               ha17d25a_0
libhwloc                  2.11.2          default_hc8275d1_1000    conda-forge
libiconv                  1.18                 h135ad9c_1    conda-forge
libjpeg-turbo             2.1.4                hcfcfb64_0    conda-forge
liblapack                 3.9.0           12_h13b7882_netlib    conda-forge
libnpp                    11.8.0.86                     0    nvidia
libnpp-dev                11.8.0.86                     0    nvidia
libnvjpeg                 11.9.0.86                     0    nvidia
libnvjpeg-dev             11.9.0.86                     0    nvidia
libpng                    1.6.39               h8cc25b3_0
libpq                     17.0                 hfc46ca6_0    conda-forge
libprotobuf               5.29.3               hac4c8cb_0
libtiff                   4.7.0                h404307b_0
libuv                     1.50.0               h2466b09_0    conda-forge
libwebp                   1.5.0                h3b0e114_0    conda-forge
libwebp-base              1.5.0                h3b0e114_0    conda-forge
libxml2                   2.13.8               h866ff63_0
lz4-c                     1.9.4                hcfcfb64_0    conda-forge
m2w64-gcc-libgfortran     5.3.0                         6    conda-forge
m2w64-gcc-libs            5.3.0                         7    conda-forge
m2w64-gcc-libs-core       5.3.0                         7    conda-forge
m2w64-gmp                 6.1.0                         2    conda-forge
m2w64-libwinpthread-git   5.0.0.4634.697f757               2    conda-forge
markupsafe                3.0.2           py310h38315fa_1    conda-forge
matplotlib-base           3.10.1          py310h37e0a56_0    conda-forge
mkl                       2023.1.0         h6a75c08_48682    conda-forge
mpmath                    1.3.0              pyhd8ed1ab_1    conda-forge
msys2-conda-epoch         20160418                      1    conda-forge
munkres                   1.1.4              pyh9f0ad1d_0    conda-forge
networkx                  3.4.2              pyh267e887_2    conda-forge
numpy                     2.2.5           py310h4987827_0    conda-forge
onnxruntime-gpu           1.21.0                   pypi_0    pypi
opencv                    4.10.0          py310h6cb580c_6
openjpeg                  2.5.2                h9b5d1b5_1
openssl                   3.5.0                ha4e3fda_1    conda-forge
openvino                  2025.1.0                 pypi_0    pypi
openvino-telemetry        2025.1.0                 pypi_0    pypi
packaging                 25.0               pyh29332c3_1    conda-forge
pandas                    2.2.3           py310hb4db72f_3    conda-forge
patsy                     1.0.1              pyhd8ed1ab_1    conda-forge
pcre2                     10.42                h0ff8eda_1
pillow                    11.1.0          py310hea0d53e_1
pip                       25.1               pyhc872135_2
pixman                    0.46.0               had0cd8c_0    conda-forge
pluggy                    1.5.0                    pypi_0    pypi
psutil                    7.0.0           py310ha8f682b_0    conda-forge
pthreads-win32            2.9.1                h2466b09_4    conda-forge
py-cpuinfo                9.0.0              pyhd8ed1ab_1    conda-forge
pycparser                 2.22               pyh29332c3_1    conda-forge
pyparsing                 3.2.3              pyhd8ed1ab_1    conda-forge
pysocks                   1.7.1              pyh09c184e_7    conda-forge
pytest                    8.3.5                    pypi_0    pypi
python                    3.10.16              h4607a30_1
python-dateutil           2.9.0.post0        pyhff2d567_1    conda-forge
python-tzdata             2025.2             pyhd8ed1ab_0    conda-forge
python_abi                3.10                    2_cp310    conda-forge
pytorch                   2.5.1           py3.10_cuda11.8_cudnn9_0    pytorch
pytorch-cuda              11.8                 h24eeafa_6    pytorch
pytorch-mutex             1.0                        cuda    pytorch
pytz                      2025.2             pyhd8ed1ab_0    conda-forge
pyyaml                    6.0.2           py310h38315fa_2    conda-forge
qhull                     2020.2               hc790b64_5    conda-forge
qtbase                    6.7.3                h0804d20_0
requests                  2.32.3             pyhd8ed1ab_1    conda-forge
scipy                     1.15.2          py310h15c175c_0    conda-forge
seaborn                   0.13.2               hd8ed1ab_3    conda-forge
seaborn-base              0.13.2             pyhd8ed1ab_3    conda-forge
setuptools                78.1.1          py310haa95532_0
six                       1.17.0             pyhd8ed1ab_0    conda-forge
sqlite                    3.45.3               h2bbff1b_0
statsmodels               0.14.4          py310hb0944cc_0    conda-forge
sympy                     1.13.1                   pypi_0    pypi
tbb                       2021.13.0            h62715c5_1    conda-forge
tk                        8.6.14               h0416ee5_0
tomli                     2.2.1                    pypi_0    pypi
torchvision               0.20.1                   pypi_0    pypi
tqdm                      4.67.1             pyhd8ed1ab_1    conda-forge
typing_extensions         4.13.2             pyh29332c3_0    conda-forge
tzdata                    2025b                h04d1e81_0
ucrt                      10.0.22621.0         h57928b3_1    conda-forge
ultralytics               8.3.127            pyh2a12c56_0    conda-forge
ultralytics-thop          2.0.14                   pypi_0    pypi
unicodedata2              16.0.0          py310ha8f682b_0    conda-forge
urllib3                   2.4.0              pyhd8ed1ab_0    conda-forge
vc                        14.42                haa95532_5
vc14_runtime              14.42.34438         hfd919c2_26    conda-forge
vs2015_runtime            14.42.34438         h7142326_26    conda-forge
wheel                     0.45.1          py310haa95532_0
win_inet_pton             1.1.0              pyh7428d3b_8    conda-forge
xz                        5.6.4                h4754444_1
yaml                      0.2.5                h8ffe710_2    conda-forge
zlib                      1.2.13               h8cc25b3_1
zstandard                 0.23.0          py310ha8f682b_2    conda-forge
zstd                      1.5.6                h8880b57_0

**Comment by UltralyticsAssistant** [comment]
👋 Hello @wenwu2021, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions may already be answered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. This should include a clear script and sample data that reproduces the difference between GPU and CPU inference in YOLO11.

If this is a custom training ❓ Question, please supply as much detail as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community wherever it suits you best! For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share and learn with others.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response. An Ultralytics engineer will also assist you soon! 😊

**Comment by Y-T-G** [comment]
It's normal behavior due to floating point erros

**Comment by wenwu2021** [comment]
Yolo11 Onnx GPU-seg c++
person:0.890330
person:0.890146
person:0.846505
bus:0.820962
Yolo11 Onnx cpu-seg c++
person:0.890330
person:0.890146
person:0.846505
bus:0.820962

Yolo11 onnxruntime c++ GPU and cpu are  same,  Yolo11 Openvino c++ GPU and cpu is  inference results are different.

**Comment by glenn-jocher** [comment]
Yes, the difference in results between GPU and CPU inference is expected behavior. This is primarily due to:

1. Floating point precision differences between hardware implementations
2. Slightly different computational approaches in different backends

While PyTorch and OpenVINO can show these minor variations, it's interesting that your ONNX C++ implementation shows consistent results across platforms. This is likely because ONNX Runtime applies more consistent computational rules regardless of hardware.

If consistent results between devices are critical for your application, you could:

1. Use a fixed random seed when possible
2. Consider quantizing your model to INT8 precision
3. Stick with the ONNX backend which seems to give you consistent results

These small differences in confidence scores shouldn't affect practical performance in most cases. The detected objects are still the same, just with slight variations in confidence values.
